tests/io/aims/test_aims_inputs.py

```diff
@@ -0,0 +1,240 @@
+
+    with pytest.raises(ValueError, match="Spin state must be 1 or 2"):
+        AimsCube(type=ALLOWED_AIMS_CUBE_TYPES[0], spinstate=3)
+
+    with pytest.raises(ValueError, match="The cube origin must have 3 components"):
+        AimsCube(type=ALLOWED_AIMS_CUBE_TYPES[0], origin=[0])
+
+    with pytest.raises(ValueError, match="Only three cube edges can be passed"):
+        AimsCube(type=ALLOWED_AIMS_CUBE_TYPES[0], edges=[[0.0, 0.0, 0.1]])
+
+    with pytest.raises(ValueError, match="Each cube edge must have 3 components"):
+        AimsCube(
+            type=ALLOWED_AIMS_CUBE_TYPES[0],
+            edges=[[0.0, 0.0, 0.1], [0.1, 0.0, 0.0], [0.1, 0.0]],
+        )
+
+    with pytest.raises(ValueError, match="elf_type only used when the cube type is elf"):
+        AimsCube(type=ALLOWED_AIMS_CUBE_TYPES[0], elf_type=1)
+
+    with pytest.raises(ValueError, match="The number of points per edge must have 3 components"):
+        AimsCube(type=ALLOWED_AIMS_CUBE_TYPES[0], points=[100, 100, 100, 100])
+
+    test_cube = AimsCube(
+        type="elf",
+        origin=[0, 0, 0],
+        edges=[[0.01, 0, 0], [0.0, 0.01, 0], [0.0, 0, 0.01]],
+        points=[100, 100, 100],
+        spinstate=1,
+        kpoint=1,
+        filename="test.cube",
+        format="cube",
+        elf_type=1,
+    )
+
+    test_cube_block = [
+        "output cube elf",
+        "    cube origin  0.000000000000e+00  0.000000000000e+00  0.000000000000e+00",
+        "    cube edge 100  1.000000000000e-02  0.000000000000e+00  0.000000000000e+00",
+        "    cube edge 100  0.000000000000e+00  1.000000000000e-02  0.000000000000e+00",
+        "    cube edge 100  0.000000000000e+00  0.000000000000e+00  1.000000000000e-02",
+        "    cube format cube",
+        "    cube spinstate 1",
+        "    cube kpoint 1",
+        "    cube filename test.cube",
+        "    cube elf_type 1",
+        "",
+    ]
+    assert test_cube.control_block == "\n".join(test_cube_block)
+
+    test_cube_from_dict = json.loads(json.dumps(test_cube.as_dict(), cls=MontyEncoder), cls=MontyDecoder)
+    assert test_cube_from_dict.control_block == test_cube.control_block
+
+
+def test_aims_control_in(tmp_path):
+    parameters = {
+        "cubes": [
+            AimsCube(type="eigenstate 1", points=[10, 10, 10]),
+            AimsCube(type="total_density", points=[10, 10, 10]),
+        ],
+        "xc": "LDA",
+        "smearing": ["fermi-dirac", 0.01],
+        "vdw_correction_hirshfeld": True,
+        "compute_forces": True,
+        "relax_geometry": ["trm", "1e-3"],
+        "batch_size_limit": 200,
+        "species_dir": str(infile_dir.parent / "species_directory/light"),
+    }
+
+    workdir = Path.cwd()
+    aims_control = AimsControlIn(parameters.copy())
+
+    for key, val in parameters.items():
+        assert aims_control[key] == val
+
+    del aims_control["xc"]
+    assert "xc" not in aims_control.parameters
+    aims_control.parameters = parameters
+
+    os.chdir(tmp_path)
```

Comment:
is the directory change necessary? maybe better to pass explicit file paths and avoid side-effects in tests? feel free to push back if you disagree.